File: 평강/app/project/src/main/java/com/eventpurchase/project/domain/product/service/ProductService.java
package com.eventpurchase.project.domain.product.service;

import com.eventpurchase.project.domain.product.dto.response.ProductResponse;
import com.eventpurchase.project.domain.product.entity.ProductEntity;
import com.eventpurchase.project.domain.product.repository.ProductRepository;
import com.eventpurchase.project.global.common.exception.CustomException;
import com.eventpurchase.project.global.common.exception.ErrorCode;
import com.eventpurchase.project.global.common.response.CursorResponse;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;


import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class ProductService {

    private final ProductRepository productRepository;

    public CursorResponse<ProductResponse> getProducts(Long lastId, int size) {

        Pageable pageable = PageRequest.of(0, size + 1);
        List<ProductEntity> products;

        if(lastId == null) {
            products = productRepository.findAll(pageable).getContent();
        }
        else{
            products = productRepository.findByIdLessThanOrderByIdDesc(lastId, pageable);
        }

        boolean hasNext = products.size() > size;
        if (hasNext) products = products.subList(0, size);
        Long nextLastId = products.isEmpty() ? null : products.getLast().getId();

        return new CursorResponse<>(products.stream().map(ProductResponse::from).toList(), nextLastId, hasNext);
    }


    public ProductResponse getProductById(Long id) {
        ProductEntity product = productRepository.findById(id)
        .orElseThrow(() -> new CustomException(ErrorCode.PRODUCT_NOT_FOUND));
        return ProductResponse.from(product);


    }

}

File: 평강/app/project/product/service/src/main/kotlin/com/eventpurchase/project/product/ProductService.kt
package com.eventpurchase.project.product

import com.eventpurchase.project.shared.response.CursorResponse
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

interface ProductService {
    fun getProducts(lastId: Long?, size: Int): CursorResponse<Product>
    fun getProductById(id: Long): Product
}

@Service
@Transactional(readOnly = true)
internal class ProductServiceImpl(
    private val productRepository: ProductRepository
) : ProductService {

    override fun getProducts(lastId: Long?, size: Int): CursorResponse<Product> {
        val products = if (lastId == null) productRepository.findAll(size + 1)
                       else productRepository.findByIdLessThan(lastId, size + 1)
        val hasNext = products.size > size
        val content = if (hasNext) products.dropLast(1) else products
        return CursorResponse(content, if (hasNext) content.last().id else null, hasNext)
    }

    override fun getProductById(id: Long): Product =
        productRepository.findById(id) ?: throw ProductNotFoundException(id)

}
